### PR TITLE
Debug pytorch tests hypersphere

### PR DESCRIPTION
**Pull request by AliceLeBrigant**
Debug of pytorch unit tests of hypersphere. The general precision of the results in the unit tests is usually a little worse in pytorch than in numpy, and so I have increases the tolerances.

Still the last one ```test_optimat_quantization.py``` is not working well, because the ```gap``` variable in the ```optimal_quantization``` method of ```riemannian_metric.py```, computed as the distance between ```center_to_update``` and ```new_center```, is sometimes zero with pytorch whereas it is non zero (but close) with numpy. I have traced this to the ```inner_product``` method of the embedding Euclidean metric, which seems to be less precise in pytorch than in numpy, but I don't understand why.

### CODE REVIEW

File: tests/test_hypersphere.py [hunk 1]
```diff
@@ -302,7 +301,7 @@ def test_squared_dist_vectorization(self):
         result = self.metric.squared_dist(n_points_a, n_points_b)
         self.assertAllClose(gs.shape(result), (n_samples, 1))
 
-    @geomstats.tests.np_and_tf_only
+    #@geomstats.tests.np_and_tf_only
```
Comment (ninamiolane) [inline tests/test_hypersphere.py]: Leftover comment

File: tests/test_hypersphere.py [hunk 2]
```diff
@@ -418,35 +417,32 @@ def test_inner_product(self):
...
         self.assertAllClose(expected, result)
 
-    @geomstats.tests.np_and_tf_only
     def test_variance(self):
         point = gs.array([0., 0., 0., 0., 1.])
-        points = gs.array([
-            point,
-            point])
+        points = gs.zeros((2, point.shape[0]))
+        points[0, :] = point
```
Comment (ninamiolane) [inline tests/test_hypersphere.py]: These assignments make the tests fail in TF. Is it possible to rewrite without assignment in ``points"?

File: tests/test_hypersphere.py [hunk 3]
```diff
@@ -418,35 +417,32 @@ def test_inner_product(self):
...
         self.assertAllClose(expected, result)
 
-    @geomstats.tests.np_and_tf_only
     def test_mean(self):
         point = gs.array([0., 0., 0., 0., 1.])
-        points = gs.array([
-            point,
-            point])
+        points = gs.zeros((2, point.shape[0]))
+        points[0, :] = point
```
Comment (ninamiolane) [inline tests/test_hypersphere.py]: Same.

File: tests/test_hypersphere.py [hunk 4]
```diff
@@ -418,35 +417,32 @@ def test_inner_product(self):
...
 
-    @geomstats.tests.np_and_tf_only
     def test_mean_and_belongs(self):
         point_a = gs.array([1., 0., 0., 0., 0.])
         point_b = gs.array([0., 1., 0., 0., 0.])
-        points = gs.array([
-            point_a,
-            point_b])
+        points = gs.zeros((2, point_a.shape[0]))
+        points[0, :] = point_a
```
Comment (ninamiolane) [inline tests/test_hypersphere.py]: Same.

File: geomstats/backend/pytorch.py [hunk 5]
```diff
@@ -296,11 +323,11 @@ def equal(a, b, **kwargs):
         a = cast(a, torch.uint8).float()
     if b.dtype == torch.ByteTensor:
         b = cast(b, torch.uint8).float()
-    return torch.equal(a, b, **kwargs)
+    return torch.eq(a, b, **kwargs)
 
 
 def floor(*args, **kwargs):
-    return torch.floor(*args, **kwargs)
+    return np.floor(*args, **kwargs)
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: Why this np.?

File: geomstats/riemannian_metric.py [hunk 6]
```diff
@@ -393,7 +393,7 @@ def optimal_quantization(self, points, n_centers=N_CENTERS,
 
         random_indices = gs.random.randint(low=0, high=n_points,
                                            size=(n_centers,))
-        centers = points[gs.ix_(random_indices, gs.arange(dimension))]
+        centers = points[gs.indexing(random_indices), :]
```
Comment (ninamiolane) [inline geomstats/riemannian_metric.py]: Could we use the function's name gs.ix_ and match numpy's API?

File: geomstats/backend/pytorch.py [hunk 7]
```diff
@@ -11,6 +11,33 @@
 int8 = 'torch.ByteTensor'
 
 
+def indexing(x):
+    return x.type(torch.LongTensor)
+
+
+def float_to_double(x):
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: Can we use the function "gs.cast" to cast into a specific type? The types can also be referred to as gs.float32, gs.float64, etc. (see backends)

For example: ```gs.cast(x, gs.float64)```

File: geomstats/backend/pytorch.py [hunk 8]
```diff
@@ -352,8 +379,8 @@ def diagonal(*args, **kwargs):
     return torch.diagonal(*args, **kwargs)
 
 
-def exp(*args, **kwargs):
-    return torch.exp(*args, **kwargs)
+def exp(input):
+    return torch.exp(torch.tensor(input))
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: Let's put the conversion to tensor outside of this function (maybe in the core code). As we do not perform such conversions for the other functions for now, it would keep the code consistent.

File: tests/test_hypersphere.py (hunk 1 above)
Comment (AliceLeBrigant) [reply tests/test_hypersphere.py]: Removed.

File: tests/test_hypersphere.py (hunk 2 above)
Comment (AliceLeBrigant) [reply tests/test_hypersphere.py]: Works now only in numpy and pytorch.

File: tests/test_hypersphere.py (hunk 3 above)
Comment (AliceLeBrigant) [reply tests/test_hypersphere.py]: Same.

File: tests/test_hypersphere.py (hunk 4 above)
Comment (AliceLeBrigant) [reply tests/test_hypersphere.py]: Same.

File: geomstats/backend/pytorch.py (hunk 5 above)
Comment (AliceLeBrigant) [reply geomstats/backend/pytorch.py]: Removed.

File: geomstats/riemannian_metric.py (hunk 6 above)
Comment (AliceLeBrigant) [reply geomstats/riemannian_metric.py]: Replaced by cast.

File: geomstats/backend/pytorch.py (hunk 7 above)
Comment (AliceLeBrigant) [reply geomstats/backend/pytorch.py]: Done.

File: geomstats/backend/pytorch.py (hunk 8 above)
Comment (AliceLeBrigant) [reply geomstats/backend/pytorch.py]: Done.
